crates/solidity/inputs/language/src/definition.rs

```diff
@@ -6627,37 +6627,366 @@ codegen_language_macros::compile!(Language(
+                ),
+                BuiltInFunction(name = "transfer", parameters = ["uint256"])
+            ]
+        ),
+        BuiltInType(
+            name = "$array",
+            fields = [BuiltInField(definition = "uint length")],
+            functions = [
+                BuiltInFunction(
+                    name = "$index",
+                    parameters = ["uint"],
+                    return_type = "$element"
+                ),
+                BuiltInFunction(
+                    name = "push",
+                    parameters = [],
+                    return_type = "$element",
+                    enabled = From("0.6.0")
+                ),
+                BuiltInFunction(
+                    name = "push",
+                    parameters = ["$element"],
+                    return_type = "uint",
+                    enabled = Till("0.6.0")
+                ),
+                BuiltInFunction(
+                    name = "push",
+                    parameters = ["$element"],
+                    enabled = From("0.6.0")
+                ),
+                BuiltInFunction(name = "pop", parameters = [])
+            ]
+        ),
+        BuiltInType(
+            name = "$arrayFixed",
+            fields = [BuiltInField(definition = "uint length")],
+            functions = [BuiltInFunction(
+                name = "$index",
+                parameters = ["uint"],
+                return_type = "$element"
+            )]
+        ),
+        BuiltInType(
+            name = "$blockType",
+            fields = [
+                BuiltInField(definition = "uint basefee", enabled = From("0.8.7")),
+                BuiltInField(definition = "uint blobbasefee", enabled = From("0.8.24")),
+                BuiltInField(definition = "uint chainid", enabled = From("0.8.0")),
+                BuiltInField(definition = "address payable coinbase"),
+                BuiltInField(definition = "uint difficulty"),
+                BuiltInField(definition = "uint gaslimit"),
+                BuiltInField(definition = "uint number"),
+                BuiltInField(definition = "uint prevrandao", enabled = From("0.8.18")),
+                BuiltInField(definition = "uint timestamp")
             ],
             functions = [BuiltInFunction(
-                name = "send",
-                parameters = ["uint256 amount"],
-                return_type = "bool"
+                name = "blockhash",
+                parameters = ["uint"],
+                return_type = "bytes32",
+                enabled = Till("0.5.0")
             )]
         ),
         BuiltInType(
-            name = "$BuiltIn$TxType",
+            name = "$bytes",
+            fields = [],
+            functions = [BuiltInFunction(
+                name = "concat",
+                parameters = ["$args"],
+                return_type = "bytes memory"
+            )]
+        ),
+        BuiltInType(
+            name = "$callOptions",
```

Comment:
moved from the limitations doc:

> The generated `built_ins.sol` file has types with a special syntax `%FooType` to prevent conflicts with user-defined types. We need to find a way to declare these types in the global scope without conflicting with users. Unfortunately structs don’t have visibility modifiers.

> We are brainstorming on a way to fix it right now, possibly trying to bind it through a another special set of rules that will add the prefixes in graph only, so that the types/names remain pure on the user side/in the CST. We can also use global functions/latest language version for all, to remove the need for the wrapper types...

Reply:
Using a second `.msgb` file only for built-ins will have the added benefit of speeding up bindings construction, since we can heavily restrict what rules need to run for built-ins. We also discussed always parsing built-ins with a fixed version of the language to avoid eg. having to put everything inside a contract because older versions of Solidity don't support top-level definitions.

The main drawback would be we'll need to maintain two separate `.msgb` files, but hopefully the scope of this second one will be a lot smaller.